src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerations/AvoidMultipleEnumerations.cs

```diff
@@ -0,0 +1,256 @@
+            if (IsDeferredType(operation.Type.OriginalDefinition, wellKnownSymbolsInfo.AdditionalDeferredTypes))
+            {
+                var isEnumerated = IsOperationEnumeratedByMethodInvocation(operation, wellKnownSymbolsInfo) || IsOperationEnumeratedByForEachLoop(operation, wellKnownSymbolsInfo);
+                if (isEnumerated)
+                    builder.Add(operation);
+            }
+        }
+
+        private void Analyze(
+            OperationBlockAnalysisContext context,
+            WellKnownTypeProvider wellKnownTypeProvider,
+            WellKnownSymbolsInfo wellKnownSymbolsInfo,
+            ImmutableHashSet<IOperation>.Builder potentialDiagnosticOperationsBuilder)
+        {
+            var potentialDiagnosticOperations = potentialDiagnosticOperationsBuilder.ToImmutable();
+            if (potentialDiagnosticOperations.IsEmpty)
+            {
+                return;
+            }
+
+            var cfg = context.OperationBlocks.GetControlFlowGraph();
+            if (cfg == null)
+            {
+                return;
+            }
+
+            // In CFG blocks there is no foreach loop related Operation, so use the
+            // the GetEnumerator method to find the foreach loop
+            var getEnumeratorSymbol = wellKnownTypeProvider
+                .GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemCollectionsGenericIEnumerable1)
+                ?.GetMembers(WellKnownMemberNames.GetEnumeratorMethodName).FirstOrDefault() as IMethodSymbol;
+
+            var analysisResult = GlobalFlowStateDictionaryAnalysis.TryGetOrComputeResult(
+                cfg,
+                context.OwningSymbol,
+                analysisContext => CreateOperationVisitor(
+                    analysisContext,
+                    wellKnownSymbolsInfo,
+                    getEnumeratorSymbol),
+                wellKnownTypeProvider,
+                context.Options,
+                MultipleEnumerableDescriptor,
+                pessimisticAnalysis: false);
+
+            if (analysisResult == null)
+            {
+                return;
+            }
+
+            var diagnosticOperations = new HashSet<IOperation>();
+            foreach (var operation in potentialDiagnosticOperations)
+            {
+                var result = analysisResult[operation.Kind, operation.Syntax];
+                if (result.Kind != GlobalFlowStateDictionaryAnalysisValueKind.Known)
+                {
+                    continue;
+                }
+
+                foreach (var kvp in result.TrackedEntities)
+                {
+                    var trackedInvocationSet = kvp.Value;
+                    if (trackedInvocationSet.EnumerationCount == InvocationCount.TwoOrMoreTime)
+                    {
+                        foreach (var trackedOperation in trackedInvocationSet.Operations)
+                        {
+                            diagnosticOperations.Add(trackedOperation);
+                        }
```

Comment:
I think we should not flag the first enumeration, only the second and subsequent enumerations. We need to order the operations by `.Syntax.SpanStart` and `.Skip(1)` to skip the lexically first enumerating operation.

Reply:
Emm, so how about the while loop case
```
while (i.First() == 1)
{
}
```
Since the first & second enumerations are the same one so we still flag this, is that correct?

I am not opposed. Just want to clarify all the cases before I make change (since it will need to modify all the tests)

Reply:
ping @mavasani  to see if you feel this is OK. Because this is the last unresolved comment in this around of code review
: )

Reply:
Hmm that does sound confusing. Seems reasonable to stay with your implementation, though I hope the message is more clear.